### DISCUSSION TITLE
ui.table with selectable rows

### DISCUSSION THREAD

**Discussion by abulka** [zauberzeug/nicegui#774] [Ideas / Feature Requests]
I'd like a ui.table with multiple columns and selectable rows, like what wxPython (a multi column [list view](https://docs.wxpython.org/wx.ListCtrl.html)):
![image](https://user-images.githubusercontent.com/11467530/232288392-fdec012f-3316-48ef-b2dc-47961bbba540.png)

the closest I could get was to simulate it using `ui.element('q-list')` and `ui.element('q-item')`

![image](https://user-images.githubusercontent.com/11467530/232288434-1bd587eb-b453-429e-b32d-15dd1029afe5.png)
The code is at https://github.com/zauberzeug/nicegui/issues/698

[falkoschindler](https://github.com/falkoschindler) wrote:
> That looks awesome! I'd love to develop a better API for it. But at the moment I'm not sure what that API might even look like. But feel free to open a new discussion about this idea/feature request. Then we can discuss it more prominently with others and eventually start implementing it.

Considerations:
- introducing proper NiceGui List and Item classes?
- model the active row so that bind_value could be used, which would presumably provide a less hacky way of finding the data of the row being clicked.
- enhance ui.table itself with this functionality?

**Comment by falkoschindler** [comment]
Thanks for the feature request, @abulka!

As we discussed on #698, we could try to build an API around q-list and q-item. The exact requirement would have to be determined.

For a `ui.table` with selectable rows - as in the title - we might be better off by fixing the "rowClick" event (https://github.com/zauberzeug/nicegui/discussions/664, https://github.com/zauberzeug/nicegui/issues/672). I just verified that we can select rows programmatically, even if the `selection` argument is `None`. We only need to fix the rowClick event.

```py
columns = [{'name': 'name', 'label': 'Name', 'field': 'name'}]
rows = [{'name': 'Alice'}, {'name': 'Bob'}, {'name': 'Carol'}]
t = ui.table(columns=columns, rows=rows, row_key='name')

def select():
    t.selected[:] = rows[:2]
    t.update()

ui.button('Select some rows', on_click=select)
```

**Comment by falkoschindler** [comment]
With the help of @bobanovo's comment https://github.com/zauberzeug/nicegui/discussions/809#discussioncomment-5696525 I just came up with this minimal example:

```py
def toggle_select(row: dict) -> None:
    if row in table.selected:
        table.selected.remove(row)
    else:
        table.selected.append(row)
    table.update()

columns = [{'name': 'name', 'label': 'Name', 'field': 'name'}]
rows = [{'name': 'Alice'}, {'name': 'Bob'}, {'name': 'Carol'}]
with ui.table(columns=columns, rows=rows, row_key='name') as table:
    table.add_slot('body-cell', r"""
        <q-td :props="props" @click="$parent.$emit('cell_click', props)">
            {{ props.value }}
        </q-td>
    """)
    table.on('cell_click', lambda msg: toggle_select(msg['args']['row']))
```

**Comment by abulka** [comment]
Am getting 
```
table.on('click', lambda msg: toggle_select(msg['args']['row']))
                                            ~~~~~~~~~~~^^^^^^^
KeyError: 'row'
```
trying to run this minimal example.

**Reply by falkoschindler** [reply]
Thanks for spotting this, @abulka!
I fixed my code by renaming the custom "click" event to "cell_click" to avoid a naming conflict with the default click event.

**Comment by abulka** [comment]
Now it works - thanks. I added some styling, multiple columns, single select logic and a display of the currently selected row.

```python
from nicegui import ui

ui.add_head_html('''
<style>
.selected {
  color: white;
  background: #59aef2; /* blue */
}
</style>
''')

multiple_selection = False
                 
def toggle_select(row: dict) -> None:
    if multiple_selection:
        if row in table.selected:
            table.selected.remove(row)
        else:
            table.selected.append(row)
    else:
        table.selected.clear()
        table.selected.append(row)
    print('toggle_select', row, table.selected)
    table.update()

columns = [{'name': 'name', 'label': 'Name', 'field': 'name'},
              {'name': 'age', 'label': 'Age', 'field': 'age'}
           ]
rows = [{'name': 'Alice', 'age': 20}, {'name': 'Bob', 'age': 30}, {'name': 'Carol', 'age': 40}]

with ui.table(columns=columns, rows=rows, row_key='name') as table:
    ui.label('hi there')
    table.add_slot('body-cell', r"""
        <q-td :props="props" @click="$parent.$emit('cell_click', props)">
            {{ props.value }}
        </q-td>
    """)
    table.on('cell_click', lambda msg: toggle_select(msg['args']['row']))

# Display the selected row, luckily bind_text_from() works
ui.label().bind_text_from(table, 'selected', lambda val: f'Current selection: {val[0]["name"] if val else "None"}')

ui.run()
```

P.S. In my electron project I found this selectable row table vue component to be pretty good too.
https://vue-smart-table.netlify.app/getting-started.html

**Reply by falkoschindler** [reply]
Oh yes, that looks nice! 😊

**Comment by meslahik** [comment]
Here is another way of selecting rows with mouse click using rowClick event, without the need for adding a slot:

```python
from nicegui import ui

columns = [
    {'name': 'name', 'label': 'Name', 'field': 'name'},
    {'name': 'age', 'label': 'Age', 'field': 'age'},
]
rows = [
    {'id': 1, 'name': 'Alice', 'age': 18},
    {'id': 2, 'name': 'Bob', 'age': 21},
    {'id': 3, 'name': 'Carol', 'age': 22},
]


def handle_row_click(e):
    """
    Source: https://quasar.dev/vue-components/table
    Quasar Table: QTable API: Events: @row-click
    """
    evt: dict = e.args[0]  # JS event object
    row: dict = e.args[1]  # The row upon which user has clicked/tapped
    row_index: int = e.args[2]  # Index of the row in the current page

    table.selected.clear()
    table.selected.append(row)

    table.update()


table = ui.table(columns=columns, rows=rows).classes('w-72')
table.on('rowClick', handle_row_click)

ui.label('').bind_text_from(table, 'selected', lambda val: f'Current selection: {val}')

ui.run()
```

**Reply by falkoschindler** [reply]
That's a great solution, @meslahik!
The click handler could be shortened to:
```py
def handle_row_click(e):
    table.selected[:] = [e.args[1]]  # The row upon which user has clicked/tapped
    table.update()
```
But a few more lines of code and comments surely don't hurt. 🙂

**Reply by bugsolarz** [reply]
@falkoschindler 
1. How to change the color of the selected row? is there any way to do it using this shortened example?
2. How to keep the table header fixed at top?

**Reply by falkoschindler** [reply]
@bugsolarz You can change the background color of the selected row with some custom CSS:
```py
ui.add_css('.q-table tr.selected { background-color: #def }')
```

And for the sticky header we can borrow the SASS code from the [Quasar documentation](https://quasar.dev/vue-components/table#example--sticky-header)
```py
ui.add_sass('''
.q-table
  .q-table__top,
  .q-table__bottom,
  thead tr:first-child th
    /* bg color is important for th; just specify one */
    background-color: white

  thead tr th
    position: sticky
    z-index: 1
  thead tr:first-child th
    top: 0

  /* this is when the loading indicator appears */
  &.q-table--loading thead tr:last-child th
    /* height of all previous header rows */
    top: 48px

  /* prevent scrolling behind sticky top row on focus */
  tbody
    /* height of all previous header rows */
    scroll-margin-top: 48px
''')
```
and add some height like `.classes('w-72 h-48')` to the table element.

**Comment by vbartle** [comment]
On this note (and happy to open another issue if there hasn't already been one) -- how does one access the [selection event ](https://www.ag-grid.com/javascript-data-grid/row-selection/#selection-events)for an aggrid table? The [listed example in docs](https://nicegui.io/documentation/ag_grid#select_ag_grid_rows) requires a button click but similar row-click functionality to the above would be great on aggrid -- is this already available and I'm just not finding it?

**Reply by falkoschindler** [reply]
You can register AG Grid events like this:
```py
grid.on('rowSelected', lambda e: print(e.args))
```
